Cabal/Distribution/Package.hs

```diff
@@ -185,12 +211,117 @@ instance NFData ComponentId where
 
 -- | Returns library name prefixed with HS, suitable for filenames
 getHSLibraryName :: UnitId -> String
-getHSLibraryName (SimpleUnitId (ComponentId s)) = "HS" ++ s
+getHSLibraryName (UnitIdVar _) = error "getHSLibraryName: unbound variable"
+getHSLibraryName uid = "HS" ++ hashUnitId uid
 
--- | For now, there is no distinction between component IDs
--- and unit IDs in Cabal.
-newtype UnitId = SimpleUnitId ComponentId
-    deriving (Generic, Read, Show, Eq, Ord, Typeable, Data, Binary, Text, NFData)
+-- | An explicit substitution on modules.
+--
+-- NB: These substitutions are NOT idempotent, for example, a
+-- valid substitution is (A -> B, B -> A).
+type ModuleSubst = Map ModuleName Module
+
+-- | Create a module substitution from a list.
+mkModSubst :: [(ModuleName, Module)] -> ModuleSubst
+mkModSubst = Map.fromList
+
+-- | Convert a module substitution into a (ascending sorted) list.
+modSubstToList :: ModuleSubst -> [(ModuleName, Module)]
+modSubstToList = Map.toAscList
+
+-- | Pretty-print the entries of a module substitution, suitable
+-- for embedding into a 'UnitId' or passing to GHC via @--instantiate-with@.
+dispModSubst :: ModuleSubst -> Disp.Doc
+dispModSubst subst
+    = Disp.hcat
+    . Disp.punctuate Disp.comma
+    $ map dispModSubstEntry (Map.toAscList subst)
+
+-- | Pretty-print a single entry of a module substitution.
+dispModSubstEntry :: (ModuleName, Module) -> Disp.Doc
+dispModSubstEntry (k, v) = disp k <<>> Disp.char '=' <<>> disp v
+
+-- | Inverse to 'dispModSubst'.
+parseModSubst :: ReadP r ModuleSubst
+parseModSubst = fmap mkModSubst
+      . flip Parse.sepBy (Parse.char ',')
+      $ parseModSubstEntry
+
+-- | Inverse to 'dispModSubstEntry'.
+parseModSubstEntry :: ReadP r (ModuleName, Module)
+parseModSubstEntry =
+    do k <- parse
+       _ <- Parse.char '='
+       v <- parse
+       return (k, v)
+
+-- | A unit identifier uniquely identifies a library (e.g.,
+-- a package) in GHC.  In the absence of Backpack, unit identifiers
+-- are just strings ('SimpleUnitId'); however, if a library is
+-- parametrized over some signatures, these identifiers need
+-- more structure.  For more details, see:
+--  https://github.com/ezyang/ghc-proposals/blob/backpack/proposals/0000-backpack.rst
+data UnitId = SimpleUnitId ComponentId
+            -- ^ Equivalent to @'UnitId' cid Map.empty@
+            | HashedUnitId ComponentId String
+            -- ^ A unit identifier where the substitution is hashed
+            | UnitId ComponentId ModuleSubst
+            | UnitIdVar !Int -- de Bruijn indexed
+    deriving (Generic, Read, Show, Eq, Ord, Typeable, Data)
+
+-- | Hash a unit identifier into a string suitable for use
+-- on a file system.
+--
+-- This function is not in "Distribution.Backpack" because we need it
+-- for 'getHSLibraryName' which has traditionally lived
+-- in "Distribution.Package". TODO: this is pretty inefficient!
+hashUnitId :: UnitId -> String
```

Comment:
Could we use a newtype for the result of displaying the unit id?

And similarly for the string used in the HashedUnitId?